src/const_eval.md

```diff
@@ -81,11 +81,127 @@ r[const-eval.const-expr.builtin-arith-logic]
-  if its lifetime is strictly contained inside the current [const context].
-  A place is *static* if it is a `static` item or a [promoted expression].
+* All forms of [borrow]s, including raw borrows, except:
+
+  * Mutable borrows of expressions whose temporary scopes would be extended (see [temporary lifetime extension]) to the end of the program.
+  * Shared borrows of expressions whose temporary scopes would be extended to the end of the program when those expressions result in values with interior mutability.
+
+  ```rust,compile_fail,E0764
+  // Due to being in tail position, this borrow extends the scope of the
+  // temporary to the end of the program. Since the borrow is mutable,
+  // this is not allowed in a const expression.
+  const C: &[u8] = &mut []; // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0764
+  // Const blocks are similar to initializers of `const` items.
+  let _: &[u8] = const { &mut [] }; // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  // This is not allowed as 1) the temporary scope is extended to the
+  // end of the program and 2) the temporary has interior mutability.
+  const C: &AtomicU8 = &AtomicU8::new(0); // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  // As above.
+  let _: &_ = const { &AtomicU8::new(0) }; // ERROR not allowed
+  ```
+
+  ```rust
+  // Even though the borrow is mutable and the temporary lives to the
+  // end of the program due to promotion, this is allowed because the
+  // borrow is not in tail position and so the scope of the temporary
+  // is not extended via temporary lifetime extension.
+  const C: () = { let _: &'static mut [u8] = &mut []; }; // OK
+  //                                              ~~
+  //                                     Promoted temporary.
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  // This shared borrow of an interior mutable temporary is allowed
+  // because its scope is not extended.
+  const C: () = { _ = &AtomicU8::new(0); }; // OK
+  ```
+
+  ```rust
+  # #![allow(static_mut_refs)]
+  // Even though this borrow is mutable, it's not of a temporary, so
+  // this is allowed.
+  const C: &u8 = unsafe { static mut S: u8 = 0; &mut S }; // OK
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  // Even though this borrow is of a value with interior mutability,
+  // it's not of a temporary, so this is allowed.
+  const C: &AtomicU8 = {
+      static S: AtomicU8 = AtomicU8::new(0); &S // OK
+  };
+  ```
+
+  > [!NOTE]
+  > In other words, shared borrows of interior mutable data and mutable borrows are only allowed in a const context when the borrowed place expression is *transient*, *indirect*, or *static*.
+  >
+  > A place expression is *transient* if it is a variable local to the current [const context] or an expression whose temporary scope is contained inside the current const context.
```

Comment:
Is this note really worth it?

Reply:
Interesting... stepped on a homonym there.  By "common", I meant "shared" rather than "frequently observed".  In the sense I meant it, putting it in the Reference (and people's use of the Reference for this purpose) makes it common.

Reply:
We're going to keep this for now.  The other note references the terminology introduced in this note, and I think it's a bit easier to explain why that odd case does what it does using this framing rather than the other framing.  But it's a bit of a coin flip.